README.md
# nodejs-repo-tools [![build status][travis_badge]][travis_link] [![coverage][coverage_badge]][coverage_link] [![Greenkeeper badge][greenkeeper_badge]][greenkeeper_link]

A tool used to maintain and test repositories in the GoogleCloudPlatform
organization.

**Table of contents**

* [Installation](#installation)
* [CLI usage](#cli-usage)
* [Programmatic usage](#programmatic-usage)
* [Language support](#language-support)
  * [Available build packs](#available-build-packs)
  * [Adding a build pack](#adding-a-build-pack)

## Installation

Via `npm`:

1.  `npm install -g @google-cloud/nodejs-repo-tools`
1.  `repo-tools --help`

Via download (Linux):

1.  `curl -O https://storage.googleapis.com/cloud-docs-samples/releases/latest/nodejs-repo-tools-linux`
1.  `mv ./nodejs-repo-tools-linux $HOME/bin/repo-tools`
1.  `chmod +x $HOME/bin/repo-tools`

Via download (Mac):

1.  `curl -O https://storage.googleapis.com/cloud-docs-samples/releases/latest/nodejs-repo-tools-macos`
1.  `mv ./nodejs-repo-tools-macos $HOME/bin/repo-tools`
1.  `chmod +x $HOME/bin/repo-tools`

Via download (Windows):

[Download link](https://storage.googleapis.com/cloud-docs-samples/releases/latest/nodejs-repo-tools-win.exe)

## CLI usage

Usage: `repo-tools --help` or `tools --help`

```
Commands:
  exec                  Run a given command in /Users/jdobry/projects/nodejs-repo-tools.
  generate <targets..>  Generate the given target(s) in /Users/jdobry/projects/nodejs-repo-tools.
  test                  Run a test sub-command.
  unify                 (Node.js only) Recursively add sub-directory dependencies to the top-level package.json file.

Options:
  --build-pack, -b  Choices: nodejs, python, ruby. Detected: nodejs. The build pack to use. The tool will attempt to
                    detect which build to use.                                                                  [string]
  --local-path, -l  Current: /Users/jdobry/projects/nodejs-repo-tools. Use this option to set the current working
                    directory of the command.                                                                   [string]
  --dry-run         Default: false. Print the actions that would be taken, but don't actually do anything.     [boolean]
  --silent          Default: false. If true, hide the output of shell commands.                                [boolean]
  --help            Show help                                                                                  [boolean]
  --version         Show version number                                                                        [boolean]

For more information, see https://github.com/GoogleCloudPlatform/nodejs-repo-tools
```

## Programmatic usage

1.  Install the tool:

        npm install --save @google-cloud/nodejs-repo-tools

1.  Then in your code:

        const tools = require('@google-cloud/nodejs-repo-tools');`

## Language support

Support for various programming languages is added via [build packs][]. A build
pack specifies language-specific config and commands that should be used when
performing the various Repo Tools tasks. Repo Tools does its best to infer the
build pack it should use, but when running a command to can specify a specific
build pack with `--build-pack [BUILD_PACK]` or `-b [BUILD_PACK]`.

### Available build packs

* [global][] - The default, global configuration for all build packs.
* [nodejs][] - Build pack for the Node.js programming language.
* [python][] - Build pack for the Python programming language.
* [ruby][] - Build pack for the Ruby programming language.

[global]: https://github.com/GoogleCloudPlatform/nodejs-repo-tools/blob/master/src/build_packs/build_pack.js
[nodejs]: https://github.com/GoogleCloudPlatform/nodejs-repo-tools/blob/master/src/build_packs/nodejs.js
[python]: https://github.com/GoogleCloudPlatform/nodejs-repo-tools/blob/master/src/build_packs/python.js
[ruby]: https://github.com/GoogleCloudPlatform/nodejs-repo-tools/blob/master/src/build_packs/ruby.js

### Adding a build pack

A build pack can be added by adding a `.js` file to the [src/build_packs][build_packs]
directory. This file should export a JavaScript object. You can see the
available options by perusing the existing build packs.

[build_packs]: https://github.com/GoogleCloudPlatform/nodejs-repo-tools/tree/master/src/build_packs

## Contributing

See [CONTRIBUTING.md](https://github.com/GoogleCloudPlatform/nodejs-repo-tools/blob/master/.github/CONTRIBUTING.md).

## License

Apache Version 2.0

See [LICENSE](https://github.com/GoogleCloudPlatform/nodejs-repo-tools/blob/master/LICENSE).

[travis_badge]: https://img.shields.io/travis/GoogleCloudPlatform/nodejs-repo-tools.svg
[travis_link]: https://travis-ci.org/GoogleCloudPlatform/nodejs-repo-tools
[coverage_badge]: https://img.shields.io/codecov/c/github/GoogleCloudPlatform/nodejs-repo-tools/master.svg
[coverage_link]: https://codecov.io/gh/GoogleCloudPlatform/nodejs-repo-tools
[greenkeeper_badge]: https://badges.greenkeeper.io/GoogleCloudPlatform/nodejs-repo-tools.svg
[greenkeeper_link]: https://greenkeeper.io/


CHANGELOG.md
##### 3.3.0 - 13 May 2019

- Update dependencies

##### 3.2.0 - 14 February 2019

- Update dependencies

##### 3.1.0 - 03 December 2018

- fix: fix: update badges on default readme (#175)
- fix: update eslint and jsdoc templates (#174)
- fix: remove contributors code (#173)

##### 3.0.0 - 09 November 2018

###### Breaking changes
- Removed the `lint` command and dependency on `semistandard`

##### 2.3.6 - 07 November 2018

###### Backwards compatible changes
- Updated dependencies
- Use new `gcloud builds` command
- Switch `semistandard` to being a non-dev dependency
- Respect URLs specified in build packs

##### 2.3.5 - 27 September 2018

###### Backwards compatible changes
- Updated dependencies

##### 2.3.3 - 27 July 2018

###### Backwards compatible changes
- Updated Readme template

##### 2.3.2 - 23 July 2018

###### Backwards compatible changes
- Added Cloud IoT Core product
- Added Cloud AutoML product

##### 2.3.1 - 11 July 2018

###### Backwards compatible changes
- Added Cloud Tasks product

##### 2.3.0 - 06 April 2018

###### Backwards compatible changes
- Added a Node.js-only "bump" command (thanks @alexander-fenster)

##### 2.2.6 - 28 March 2018

- Add Text-to-Speech product info

###### Bug fixes
- Fix link to open in cloud shell button image

##### 2.2.5 - 22 March 2018

- Added comments in Node.js README templates that the files are autogenerated.

##### 2.2.4 - 16 March 2018

###### Bug fixes
- Add missing lines to `.gitignore` Node.js template
- Add missing configuration for two products

##### 2.2.3 - 14 March 2018

###### Bug fixes
- Fix link to "Open in Cloud Shell" button image

##### 2.2.2 - 13 March 2018

###### Bug fixes
- Un-ignore package-lock.json files in Node.js .gitignore template

##### 2.2.1 - 21 February 2018

###### Bug fixes
- Removed broken links from Node.js issue template

##### 2.2.0 - 26 January 2018

###### Backwards compatible changes
- Improved Python support
- Upgraded dependencies

##### 2.1.4 - 17 January 2018

- Updated a product name

##### 2.1.3 - 17 November 2017

###### Backwards compatible changes
- #81 - Add ability to specify test url in config file

##### 2.1.2 - 08 November 2017

###### Backwards compatible changes
- Added `yaml` flag to the `test deploy` command.

##### 2.1.1 - 01 November 2017

###### Bug fixes
- Fix issue with `getUrl`

##### 2.1.0 - 27 October 2017

###### Bug fixes
- #59 - Link to license and contributing files in lib_readme are wrong.
- #60 - Add link to the library's repo in its readme
- #61 - Compatible when sample directory does not exist
- #67 - Samples link broken in the docs

###### Backwards compatible changes
- #56 - Successfully ignore the new package-lock.json-12345 files.
- #63 - Add --except option to generate command to be used in combination with the "all" target
- #65 - Support deprecated release level.
- Add Open in Cloud Shell button to readmes.
- Dogfood repo-tools on repo-tools. Switched to eslint + prettier.

###### Other
- Updated dependencies

##### 2.0.11 - 22 October 2017

###### Bug fixes
- Fix to `spawnAsyncWithIO`

##### 2.0.10 - 17 October 2017

###### Backwards compatible changes
- Fixed wrong samples link in lib_readme.tpl
- #52 - Make alpha release level badges orange.
- #54 - Remove the autogen beta disclaimer.

###### Other
- Upgraded dependencies

##### 2.0.9 - 12 October 2017

###### Backwards compatible changes
- Updated eslintrc_systemtest.tpl

##### 2.0.8 - 12 October 2017

###### Backwards compatible changes
- Added support for generating and updating a package.json file.
- Added support for generating a CONTRIBUTORS file.

##### 2.0.7 - 03 October 2017

###### Bug fixes
- Fix Firestore docs link

##### 2.0.6 - 03 October 2017

###### Backwards compatible changes
- #43 - Fix "quickstart" capitalization
- #45 - Add `.prettierignore`
- Use official Apache 2.0 license template.
- Provide a mechanism to suppress the billing instruction.
- Add Cloud Firestore product information.
- #46 - Codecov badges should use master branch.

##### 2.0.5 - 02 October 2017

###### Other
- Tweaked library readme
- Fix for JSDoc config

##### 2.0.4 - 22 September 2017

###### Backwards compatible changes
- #39 - Add ability to hide "Enable API" if api_id doesn't exist
- #40 - Add all API IDs to src/utils/products.js

##### 2.0.3 - 22 September 2017

###### Bug fixes
- Fixed --help and --version flags

##### 2.0.2 - 22 September 2017

###### Backwards compatible changes
- Made command logs more consistent

##### 2.0.1 - 22 September 2017

###### Backwards compatible changes
- #36 - Automatically write a JavaScript appropriate .gitignore

###### Bug fixes
- Fixed release script

###### Other
- Updated contributing guide

##### 2.0.0 - 22 September 2017

Stable 2.0.0 release

###### Bug fixes
- #35 - Fix logger stutter

##### 2.0.0-beta.12 - 20 September 2017

###### Breaking changes
- Changed signature of `lint` command to match that of the `test` commands, i.e. using `--cmd` and `--` for args.

###### Other
- Improved repository URL handling
- Upgraded dependencies

##### 2.0.0-beta.11 - 19 September 2017

- Small tweak to the `lib_readme` template.

##### 2.0.0-beta.10 - 18 September 2017

- Fixed typo in `lib_readme` template.

##### 2.0.0-beta.9 - 18 September 2017

- Make `lib_readme` quickstart sample region tag configurable.
- Improve the look of the samples list in `lib_readme`.

##### 2.0.0-beta.8 - 18 September 2017

- Updates to `templates/nodejs/jsdoc.tpl`

##### 2.0.0-beta.7 - 07 September 2017

- Added 6 new targets to the `generate` command.
- Added `shell: true` to improve stability of CLI commands.

##### 2.0.0-beta.6 - 29 August 2017

- Added `templates/` to distribution (it was missing)

##### 2.0.0-beta.5 - 29 August 2017

- Fix Release Quality badge in lib_readme
- #23 - Add syntax highlighting to quickstart sample in lib_readme
- Fix relative Github links in lib_readme
- #22 - Add support for year ranges in the license file.

##### 2.0.0-beta.4 - 23 August 2017


src/utils/index.js
/**
 * Copyright 2018, Google LLC.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';

require('colors');

const childProcess = require('child_process');
const got = require('got');
const net = require('net');
const url = require('url');

const {spawn} = require('child_process');

const MAX_TRIES = 8;

exports.parseArgs = (args = '') => {
  if (Array.isArray(args)) {
    return args;
  }

  const parsed = [];
  let arg = '';
  let inQuote = false;
  let quoteChar = null;

  for (let i = 0; i < args.length; i++) {
    if (args[i] === `'`) {
      if (inQuote) {
        if (args[i - 1] === `\\`) {
          arg += args[i];
        } else {
          if (quoteChar === `'`) {
            arg += args[i];
            inQuote = false;
            quoteChar = null;
          } else {
            arg += args[i];
          }
        }
      } else {
        if (args[i - 1] === `\\`) {
          arg += args[i];
        } else {
          arg += args[i];
          quoteChar = `'`;
          inQuote = true;
        }
      }
    } else if (args[i] === `"`) {
      if (inQuote) {
        if (args[i - 1] === `\\`) {
          arg += args[i];
        } else {
          if (quoteChar === `"`) {
            arg += args[i];
            inQuote = false;
            quoteChar = null;
          } else {
            arg += args[i];
          }
        }
      } else {
        if (args[i - 1] === `\\`) {
          arg += args[i];
        } else {
          arg += args[i];
          quoteChar = `"`;
          inQuote = true;
        }
      }
    } else if (/\s/.test(args[i])) {
      if (inQuote) {
        arg += args[i];
      } else if (arg) {
        parsed.push(arg);
        arg = ``;
      } else {
        args = ``;
      }
    } else {
      arg += args[i];
    }

    if (i === args.length - 1) {
      if (inQuote) {
        throw new Error(`Unclosed quote in: ${args}`);
      } else if (arg) {
        parsed.push(arg);
      }
    }
  }

  return parsed;
};

exports.getUrl = opts => {
  return `https://${opts.version}-dot-${opts.project}.appspot.com`;
};

exports.finalize = (err, resolve, reject) => {
  if (err) {
    reject(err);
  } else {
    resolve();
  }
};

const logger = {
  error(config, ...args) {
    // eslint-disable-next-line no-console
    console.error(
      `${(typeof config === 'string' ? config : config.test).bold.red}:`,
      ...args.map(arg => {
        if (arg.red) {
          return arg.red;
        }
        return arg;
      })
    );
  },

  fatal(...args) {
    this.error(...args);
    // eslint-disable-next-line no-process-exit
    process.exit(1);
  },

  log(config, ...args) {
    // eslint-disable-next-line no-console
    console.log(
      `${(typeof config === 'string' ? config : config.test).bold}:`,
      ...args
    );
  },

  debug(...args) {
    if (process.argv.indexOf('--debug') !== -1) {
      // eslint-disable-next-line no-console
      console.log('debug'.bold, ...args);
    }
  },
};

exports.logger = logger;

exports.getRepoPath = (repository, cwd) => {
  repository || (repository = {});
  if (typeof repository === 'string') {
    repository = {
      url: repository,
    };
  }

  if (!repository.url) {
    let pushUrl = childProcess
      .execSync('git remote get-url --push origin', {
        cwd,
        stdout: 'silent',
      })
      .toString()
      .trim();
    const start = pushUrl.indexOf('github.com');
    if (start >= 0) {
      pushUrl = pushUrl.substring(start + 11);
      if (pushUrl) {
        return `/${pushUrl.replace('.git', '')}`;
      }
    }
  }

  const result = url.parse(repository.url).path.replace('.git', ''); // eslint-disable-line node/no-deprecated-api
  if (!result.startsWith('/')) {
    return `/${result}`;
  }
  return result;
};

exports.getTimeTaken = start => {
  let timeTaken = (Date.now() - start) / 1000;
  if (timeTaken <= 100) {
    timeTaken = timeTaken.toPrecision(3);
  } else if (timeTaken >= 100) {
    timeTaken = Math.floor(timeTaken);
  }
  return `${timeTaken}s`.cyan;
};

/**
 * Generates a markdown badge for displaying a "Release Quality'.
 *
 * @param {string} releaseQuality One of: (ga, beta, alpha, eap, deprecated).
 * @returns {string} The markdown badge.
 */
exports.createReleaseQualityBadge = releaseQuality => {
  releaseQuality = releaseQuality.toUpperCase();
  let badge = '';
  if (releaseQuality === 'GA') {
    badge = 'general%20availability%20%28GA%29-brightgreen';
  } else if (releaseQuality === 'BETA') {
    badge = 'beta-yellow';
  } else if (releaseQuality === 'ALPHA') {
    badge = 'alpha-orange';
  } else if (releaseQuality === 'EAP') {
    badge = 'EAP-yellow';
  } else if (releaseQuality === 'DEPRECATED') {
    badge = 'deprecated-red';
  } else {
    logger.error(
      'generate',
      `Expected "release_quality" to be one of: (ga, beta, alpha, eap, deprecated)! Actual: "${releaseQuality}"`
    );
    // eslint-disable-next-line no-process-exit
    process.exit(1);
  }
  return `[![release level](https://img.shields.io/badge/release%20level-${badge}.svg?style=flat)](https://cloud.google.com/terms/launch-stages)`;
};

let portrange = 45032;
const triedPorts = {};

exports.getPort = config => {
  let port = config.port || portrange;

  while (triedPorts[port]) {
    port += 1;
  }

  triedPorts[port] = true;

  return new Promise(resolve => {
    const server = net.createServer();
    server.listen(port, () => {
      server.once('close', () => {
        resolve(port);
      });
      server.close();
    });
    server.on('error', () => {
      resolve(exports.getPort(config));
    });
  });
};

exports.testRequest = (url, config, numTry) => {
  logger.log('app', `Verifying: ${url.yellow}`);
  if (!numTry) {
    numTry = 1;
  }

  let canTryAgain = false;

  return got(url, {
    timeout: 10000,
  })
    .then(
      response => {
        const EXPECTED_STATUS_CODE = config.code || 200;

        const body = response.body || '';
        const code = response.statusCode;

        if (code !== EXPECTED_STATUS_CODE) {
          canTryAgain = true;
          throw new Error(
            `failed verification!\nExpected status code: ${EXPECTED_STATUS_CODE}\nActual: ${code}`
          );
        } else if (config.msg && !body.includes(config.msg)) {
          throw new Error(
            `failed verification!\nExpected body: ${
              config.msg
            }\nActual: ${body}`
          );
        } else if (config.testStr && !config.testStr.test(body)) {
          throw new Error(
            `failed verification!\nExpected body: ${
              config.testStr
            }\nActual: ${body}`
          );
        }
      },
      err => {
        canTryAgain = true;
        if (err && err.response) {
          const EXPECTED_STATUS_CODE = config.code || 200;

          const body = err.response.body || '';
          const code = err.response.statusCode;

          if (code !== EXPECTED_STATUS_CODE) {
            throw new Error(
              `failed verification!\nExpected status code: ${EXPECTED_STATUS_CODE}\nActual: ${code}`
            );
          } else if (config.msg && !body.includes(config.msg)) {
            throw new Error(
              `failed verification!\nExpected body: ${
                config.msg
              }\nActual: ${body}`
            );
          } else if (config.testStr && !config.testStr.test(body)) {
            throw new Error(
              `failed verification!\nExpected body: ${
                config.testStr
              }\nActual: ${body}`
            );
          }
        } else {
          return Promise.reject(err);
        }
      }
    )
    .catch(err => {
      if (numTry >= MAX_TRIES || !canTryAgain) {
        return Promise.reject(err);
      }

      return new Promise((resolve, reject) => {
        setTimeout(() => {
          exports.testRequest(url, config, numTry + 1).then(resolve, reject);
        }, 500 * Math.pow(numTry, 2));
      });
    });
};

// Delete an App Engine version
exports.deleteVersion = config => {
  return new Promise((resolve, reject) => {
    const cmd = config.deleteCmd || 'gcloud';

    logger.log(config, 'Deleting deployment...');
    // Keep track off whether "done" has been called yet
    let calledDone = false;

    const args = [
      `app`,
      `versions`,
      `delete`,
      config.test,
      `--project=${config.projectId}`,
      `-q`,
    ];

    const child = spawn(cmd, args, {
      cwd: config.cwd,
      // Shouldn't take more than 4 minutes to delete a deployed version
      timeout: 4 * 60 * 1000,
    });

    logger.log(
      config,
      `Delete command: ${(cmd + ' ' + args.join(' ')).yellow}`
    );

    child.on('error', finish);

    child.stdout.on('data', data => {
      const str = data.toString();
      if (str.includes('\n')) {
        process.stdout.write(`${config.test.bold}: ${str}`);
      } else {
        process.stdout.write(str);
      }
    });
    child.stderr.on('data', data => {
      const str = data.toString();
      if (str.includes('\n')) {
        process.stderr.write(`${config.test.bold}: ${str}`);
      } else {
        process.stderr.write(str);
      }
    });

    child.on('exit', code => {
      if (code !== 0) {
        finish(new Error(`${config.test}: failed to delete deployment!`));
      } else {
        finish();
      }
    });

    // Exit helper so we don't call "cb" more than once
    function finish(err) {
      if (!calledDone) {
        calledDone = true;
        exports.finalize(err, resolve, reject);
      }
    }
  });
};
